### ISSUE TITLE
Does CNTK support multiple output layers?

### ISSUE THREAD

**Issue by lofrank** [Microsoft/CNTK#510]
I'm trying to use CNTK to implement Multi-digit Number Recognition from Street View Imagery using Deep Convolutional Neural Networks (http://arxiv.org/pdf/1312.6082.pdf).

In Torch, we can add more than one output layers (each for one character here), like this:
_--[[SEQUENCER - classifies the length and all digits ]]-- 
sequencer = nn.**ConcatTable**() 
sequencer:add(lengthPredictor) 
sequencer:add(classifier(lengthClasses)) -- length predictor 
for i = 1, maxDigits do 
        sequencer:add(classifier(digitClasses)) -- digit class predictor 
end 
--[[MODEL]]-- 
model = nn.Sequential() 
model:add(extractor) -- H 
model:add(sequencer)_

This is helpful when the output is a sequence of characters (In MNSIT example, there is only one character in output).

I have go through the wiki, and did not find how to do this, anyone can give some comments?

Thanks.

**Comment by frankseide** [comment]
Yes. In CNTK, the criterion is a user-defined expression, which gets summed up over all samples of a minibatch. If I understand the paper right, then
```
sample-level criterion  = length criterion + sum over all digit-level criteria
```
In CNTK, this would look like something like this:
```
ce = CrossEntropyWithSoftmax (trueLen, lenLogP)
         + CrossEntropyWithSoftmax (trueDigit1, digit1LogP)
         + ...
         + CrossEntropyWithSoftmax (trueDigitN, digitNLogP)
```
assuming you use a per-digit Softmax for each. (This is a sketch, we can do better, e.g. the 'logP' values do not need to be normalized, and there is a way of writing this more concisely with a loop, let us look at that later.)

We would need to extend the above to deal with varying lengths; you would want to exclude summands for any digit above the actual length. This can be done by multiplying the respective summand with 0/1 flag that is 0 for those positions that have no digit.

BTW We are revamping our description language. The above requires the new "BrainScriptNetworkBuilder." The documentation for that will come online within days, stay tuned.

**Comment by lofrank** [comment]
@frankseide 
Thanks for your quick response. 
It's happy to know that I can get summed up over all the CrossEntropyWithSoftmax. 
All the examples that using UCIFastReader has dim = 1 for labels, would you please tell me how to deal labels when dim of labels is greater than 1?

In 01_OneHidden.ndl, variable named 'labels' is used as the first parameter of the CrossEntropyWithSoftmax function.
_labels = InputValue(labelDim)
ce = CrossEntropyWithSoftmax(labels, ol)
...
LabelNodes = (labels)_

Assume that the length of the output sequence is fixed to 4, should I define for variables like this? If yes, how do I assign more than one value to **LableNodes**?
_trueDigit1 = InputValue(labelDim)
trueDigit2 = InputValue(labelDim)
trueDigit3 = InputValue(labelDim)
trueDigit4 = InputValue(labelDim)_


The NDL is not so intuitive for me, the worse is that only three of the NDL functions have detailed definitions.
https://github.com/Microsoft/CNTK/wiki/Full-NDL-Function-Reference

I will be appreciate if you can implement an example based on the 01_OneHidden.ndl under Image\MSNIST\Config folder.

Thanks again.

**Comment by lofrank** [comment]
With CNTKTextFormatReader, I can define more labels.
But still don't know how to define multiple output layers...

**Comment by gaizkan** [comment]
Hi lofrank,
I am not 100% certain I understand your question. Do you want to specify which nodes are to be considered as output nodes?
You can specify the output nodes with the OutputNodes property.

e.g.: 
    `
    OutputNodes="h1.z:ol.z"
    `
Use colon to separate them.
Does this answer your question?

Regards,

Gaizka
